### PR TITLE
ConjoinXRuns can receive input workspaces containing Dx values

### LINKED ISSUE #22218: Algorithm ConjoinXRuns does not consider Dx values

### PR DESCRIPTION
**Pull request by vreimund**
ConjoinXRuns can now join workspaces which contain Dx values.

- Update of RunCombinationHelper in order to account for Dx value check per spectra. Note this happens during input validation and increases execution time. (used by MergeRuns).

- Add optional Dx to WorkspaceCreationHelper methods
  - Do not overload create2DWorkspaceBinned any more in WorkspaceCreationHelper

- No update of the documentation in terms of usage examples since this functionality (like masking or error propagation) are normal and no highlight of this algorithm. 

Commit messages contain information.

**To test:**

All tests should pass, added unit tests.

Maybe use CreateWorkspace or use CreateSampleWorkspace, add Dx values.

For example:
```Python
values = [1., 2., 3.]
ws = CreateWorkspace(DataX = values, DataY = values, Dx = values)
new_values = [4., 5., 6.]
ws2 = CreateWorkspace(DataX = new_values, DataY = new_values, Dx = new_values)
ws3= ConjoinXRuns([ws, ws2])
```

Fixes #22218 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
A couple of comments:

File: Framework/Algorithms/src/ConjoinXRuns.cpp [hunk 1]
```diff
@@ -271,28 +275,32 @@ void ConjoinXRuns::fillHistory() {
...
+      auto dx = input->dx(index);
+      xerrors.insert(xerrors.end(), dx.begin(), dx.end());
     }
-    spectrum.insert(spectrum.end(), y.begin(), y.end());
-    errors.insert(errors.end(), e.begin(), e.end());
-    axis.insert(axis.end(), x.begin(), x.end());
   }
-
+  if (!xerrors.empty()) {
+    m_outWS->setPointStandardDeviations(index, m_outWS->blocksize());
```
[the remark this answers is not in the document]
Comment (reimundILL) [reply Framework/Algorithms/src/ConjoinXRuns.cpp]: That would be nicer indeed, I will do this

File: Framework/Algorithms/src/ConjoinXRuns.cpp [hunk 2]
```diff
@@ -271,28 +275,32 @@ void ConjoinXRuns::fillHistory() {
...
   spectrum.reserve(m_outWS->blocksize());
   errors.reserve(m_outWS->blocksize());
   axis.reserve(m_outWS->blocksize());
   size_t index = static_cast<size_t>(wsIndex);
-
   for (const auto &input : m_inputWS) {
-    auto y = input->y(index).rawData();
-    auto e = input->e(index).rawData();
-    std::vector<double> x;
+    HistogramData::HistogramY y = input->y(index);
```
[the remark this answers is not in the document]
Comment (reimundILL) [reply Framework/Algorithms/src/ConjoinXRuns.cpp]: Will change it.

File: Framework/Algorithms/src/ConjoinXRuns.cpp [hunk 3]
```diff
@@ -271,28 +275,32 @@ void ConjoinXRuns::fillHistory() {
 * @param wsIndex : the workspace index
 */
 void ConjoinXRuns::joinSpectrum(int64_t wsIndex) {
-  std::vector<double> spectrum;
-  std::vector<double> errors;
-  std::vector<double> axis;
+  std::vector<double> spectrum, errors, axis, x, xerrors;
```
[the remark this answers is not in the document]
Comment (reimundILL) [reply Framework/Algorithms/src/ConjoinXRuns.cpp]: Will change this back.

**Comment by SimonHeybrock** [comment]
Changes look good now, tests were added. :+1:

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**